compiler/rustc_type_ir/src/infer_ctxt.rs

```diff
@@ -1,9 +1,46 @@
+use derive_where::derive_where;
+#[cfg(feature = "nightly")]
+use rustc_macros::{HashStable_NoContext, TyDecodable, TyEncodable};
+
 use crate::fold::TypeFoldable;
+use crate::inherent::*;
 use crate::relate::RelateResult;
 use crate::relate::combine::PredicateEmittingRelation;
-use crate::solve::SolverMode;
+use crate::solve::Reveal;
 use crate::{self as ty, Interner};
 
+#[derive_where(Clone; I: Interner)]
+#[derive_where(Hash; I: Interner)]
+#[derive_where(PartialEq; I: Interner)]
+#[derive_where(Eq; I: Interner)]
+#[derive_where(Debug; I: Interner)]
+#[derive_where(Copy; I: Interner)]
+#[cfg_attr(feature = "nightly", derive(TyEncodable, TyDecodable, HashStable_NoContext))]
+pub enum TypingMode<I: Interner> {
+    Coherence,
+    Analysis { defining_opaque_types: I::DefiningOpaqueTypes },
+    PostAnalysis,
+}
+
+impl<I: Interner> TypingMode<I> {
+    pub fn non_body_analysis() -> TypingMode<I> {
+        TypingMode::Analysis { defining_opaque_types: Default::default() }
+    }
+
+    pub fn analysis_in_body(cx: I, body_def_id: I::LocalDefId) -> TypingMode<I> {
+        TypingMode::Analysis { defining_opaque_types: cx.opaque_types_defined_by(body_def_id) }
+    }
+
+    // FIXME(@lcnr): Using this function is questionable as the `param_env`
```

Comment:
Yeah, eventually we should pass typing mode around as the primary piece of, or some sort of pair between `TypingModeAndParamEnv` which `infcx.build()` splits and will give you back an `InferCtxt, ParamEnv` pair.

Reply:
Also, all of these `FIXME(@lcnr)` would ideally be changed into `FIXME(typing_mode)` so we can track this effort more specifically?